llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -6405,7 +6405,8 @@ SDValue TargetLowering::BuildSDIV(SDNode *N, SelectionDAG &DAG,
     if (VT.isVector())
       WideVT = EVT::getVectorVT(*DAG.getContext(), WideVT,
                                 VT.getVectorElementCount());
-    if (isOperationLegalOrCustom(ISD::MUL, WideVT)) {
+    if (isOperationLegalOrCustom(ISD::MUL, WideVT) ||
```

Comment:
I think it would make more sense to just delete the check

Reply:
I tried enabling it for all targets but it doesn't seem to always be profitable. At minimum we need a isTypeLegal check otherwise some targets even crash (riscv32).

I'll have another look

Reply:
Removing the check causes about 30 failures. Using only isTypeLegal on the wide type doesn't cause more than 1-2 test changes but then it doesn't work on AMDGPU because i128 isn't legal

Reply:
I primarily want to not have any new TLI hook. Quantity of test changes doesn't mean it's worse, just update them?

Reply:
I assume this should be something like !LegalOperations || isOperationLegalOrCustom(MUL)

Reply:
Sure, I can update the tests, but we still need some kind of legality check, otherwise we get crashes in a few places. RISCV seems especially sensitive to that and there's a few crashes I can't seem to get rid of without getting rid of the AMDGPU improvements.

Why is the TLI hook not desirable? We already have some for other expansions. 
I could make an enum instead to decide which expansion method to use, and targets would default to a legal method while AMDGPU can force MUL expansion.

Reply:
> Why is the TLI hook not desirable?

Because we already have way too many garbage single use, undiscoverable hacks for specific combines like this one. 

> We already have some for other expansions.

That's the problem. They're all hacky shortcuts to avoid touching all the targets or coming up with a reasonable heuristic in terms of the existing APIs